### PR TITLE
Add build for arm64

### PR DESCRIPTION
**Pull request by unkcpz**

### CODE REVIEW

File: Dockerfile
Comment (ltalirz) [inline Dockerfile]: Why is this needed now?

File: Dockerfile
Comment (ltalirz) [inline Dockerfile]: @yakutovicha Are you ok with deleting/commenting this (for me it's fine)?

If not, I suggest you just put an if/else switch here based on the `ARCH` (as for miniconda below).
Note that you will need to add the mirror for the correct ubuntu version (it's not longer bionic)

**Review by ltalirz — CHANGES_REQUESTED** [review]
thanks @unkcpz , nice work!

I have two minor comments, one for @yakutovicha

File: Dockerfile
Comment (unkcpz) [reply Dockerfile]: No idea why it is needed for arm64 only, without this will cause a problem of running aiida-core docker image but not this image. It is header files for PostgreSQL I think.

Comment (ltalirz) [reply Dockerfile]: If you still remember what this is needed for in aiida-core, it might be worth writing it in a comment here

File: Dockerfile
Comment (unkcpz) [inline Dockerfile:75]: The arch name is aarch64 rather than `arm64` by `uname -m`. A minor update I added here.

File: Dockerfile
Comment (unkcpz) [reply Dockerfile]: Add a comment for this, not pretty sure I interpret the error correctly I paste below the error of docker build aiida-core without this `libpq-dev`. 

```
#7 51.98     gcc -pthread -B /opt/conda/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC "-DPSYCOPG_VERSION=2.8.6 (dt dec pq3 ext lo64)" -DPG_VERSION_NUM=130005 -DHAVE_LO64=1 -I/opt/co
nda/include/python3.8 -I. -I/usr/include/postgresql -I/usr/include/postgresql/13/server -c psycopg/adapter_asis.c -o build/temp.linux-aarch64-3.8/psycopg/adapter_asis.o -Wdeclaration-after-statement
#7 51.98     In file included from psycopg/adapter_asis.c:28:
#7 51.98     ./psycopg/psycopg.h:36:10: fatal error: libpq-fe.h: No such file or directory
#7 51.98        36 | #include <libpq-fe.h>
#7 51.98           |          ^~~~~~~~~~~~
#7 51.98     compilation terminated.
#7 51.98     
#7 51.98     It appears you are missing some prerequisite to build the package from source.
#7 51.98     
#7 51.98     You may install a binary package by installing 'psycopg2-binary' from PyPI.
#7 51.98     If you want to install psycopg2 from source, please install the packages
#7 51.98     required for the build and try again.
#7 51.98     
#7 51.98     For further information please check the 'doc/src/install.rst' file (also at
#7 51.98     <https://www.psycopg.org/docs/install.html>).
#7 51.98     
#7 51.98     error: command '/usr/bin/gcc' failed with exit code 1
#7 51.98     ----------------------------------------
#7 51.98 ERROR: Command errored out with exit status 1: /opt/conda/bin/python3.8 -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-825i6nvd/psycopg2-binary_04de1ed0a32a40df8a7edf5b4596a4fe/setup.py'"'"';
 __file__='"'"'/tmp/pip-install-825i6nvd/psycopg2-binary_04de1ed0a32a40df8a7edf5b4596a4fe/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; 
setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-ihv_tvxr/install-record.txt --single-version-externally-managed --compile --inst
all-headers /opt/conda/include/python3.8/psycopg2-binary Check the logs for full command output.
------
error: failed to solve: executor failed running [/bin/sh -c pip install ./aiida-core[atomic_tools]]: exit code: 1
```

Comment (ltalirz) [reply Dockerfile]: I think you interpreted correctly - it makes sense that they don't yet have wheels for that architecture

File: .github/workflows/push_to_dockerhub.yml
Comment (yakutovicha) [inline .github/workflows/push_to_dockerhub.yml:51]: Can we also run tests on these architectures?

File: Dockerfile
Comment (yakutovicha) [inline Dockerfile]: Both `impish` and `hirsute` are GitHub branches, not tags. They might be updated in the future, which is not what we necessarily want. But at the same time, we want to support the arm architectures.

Very unfortunate situation

File: Dockerfile
Comment (yakutovicha) [reply Dockerfile]: I am definitely ok with it, as soon as the issue https://github.com/aiidalab/aiidalab-docker-stack/issues/9 does not pop up again. But since we are building things on Docker hub - it shouldn't.

**Review by yakutovicha — COMMENTED** [review]
thanks, @unkcpz, I left a couple of comments below.

Overall (not related to this PR), I am a bit concerned about the phusion/base-image. I am not sure the issues we encounter with using it (very inefficient support, long delays for releasing new things) outweigh the advantages. Maybe we should just go with the official ubuntu docker images.

File: Dockerfile
Comment (ltalirz) [reply Dockerfile]: If I understand correctly, this is simply upgrading to ubuntu 21.10
https://github.com/phusion/baseimage-docker/commit/cf8e398043707f2a102a17090c927ac6ef3d0ed2

Is this necessary for the arm support?
If not, we can stick with focal-1.1.0

**Comment by ltalirz** [comment]
@unkcpz do you plan finishing up this PR?

File: Dockerfile
Comment (unkcpz) [reply Dockerfile]: sadly, this is mandatory to support arm.

**Comment by unkcpz** [comment]
> @unkcpz do you plan finishing up this PR?

I think we can leave this open or change it to an open issue. If I understand it correctly, @yakutovicha mentioned in aiidalab meeting that he has a plan to use some more wild use base image instead of `phusion`. 

@yakutovicha could you elaborate more about the plan and write down more details about why we need `phusion` at first place and why now it is possible to replace it?

**Comment by ltalirz** [comment]
> @yakutovicha could you elaborate more about the plan and write down more details about why we need `phusion` at first place and why now it is possible to replace it?

The original reason for using the phusion baseimage was that it [provides an init service](https://blog.phusion.nl/2015/01/20/docker-and-the-pid-1-zombie-reaping-problem/) which is designed for the use case we have (running multiple processes inside a docker container) and can in principle prevent containers from becoming filled with zombie processes.
However, zombie processes arise from bugs in application code, and to my knowledge we have never actually encountered the problem ourselves before switching (correct me if I'm wrong).

It also runs the `cron` daemon by default (not sure whether we're taking advantage of this).

I personally don't have a strong opinion in either direction - we should just take a decision.

**Comment by ltalirz** [comment]
Just mentioning that we may soon be preparing a small update of aiida-prerequisites in order to fix the rabbitmq configuration.

Would be good if you guys can get this sorted and merged before to have the arm support as well.

**Comment by yakutovicha** [comment]
> Just mentioning that we may soon be preparing a small update of aiida-prerequisites in order to fix the rabbitmq configuration.
> 
> Would be good if you guys can get this sorted and merged before to have the arm support as well.

Hi @ltalirz - I've been thinking for quite a while about a better approach to handle containers. I came up with this: https://github.com/aiidalab/aiidalab-docker-stack/issues/243.

In principle, I am fully into it now. If I get positive feedback and green light from the others - I can replace the thing pretty quickly. As the result, there won't be the prerequisites container. Let me know if you like the idea.

**Comment by yakutovicha** [comment]
@unkcpz let's try to finalise this PR. The 22.04 Ubuntu based image [has been released](https://github.com/phusion/baseimage-docker/releases/tag/jammy-1.0.0) 20 days ago, so let's try to finalise the PR. Would you have time to do the remaining changes?

**Comment by unkcpz** [comment]
@yakutovicha Thanks for head up. No problem, I think there is no blocker and issues with this implementation.  I will give it a test on my local machine again, rebase commits and let CI build test run once more.

**Comment by unkcpz** [comment]
I rebase the PR and update miniconda version. It can be built and for aiida-core v1.6.8 container and then for aiidalab-docker-stack. But when I launch it I get errors below. I have no idea how to fix this. For the CI build test failed here https://github.com/aiidateam/aiida-prerequisites/runs/6789039113?check_suite_focus=true. I remove line `RUN touch /opt/conda/pkgs/urls.txt`. Not sure if this cause the issue? The rabbitmq inside aiida-prerequisetes container is not installed by conda.

```
*** Running /etc/my_init.d/10_syslog-ng.init...
[2022-06-08T09:05:50.693268] WARNING: Configuration file format is too old, syslog-ng is running in compatibility mode. Please update it to use the syslog-ng 3.35 format at your time of convenience. To upgrade the configuration, please review the warnings about incompatible changes printed by syslog-ng, and once completed change the @version header at the top of the configuration file; config-version='3.25'
[2022-06-08T09:05:50.787845] WARNING: The internal_queue_length stat counter has been renamed to internal_source.queued. The old name will be removed in future versions; config-version='3.25'
Jun  8 09:05:50 07d54204d21c syslog-ng[476]: syslog-ng starting up; version='3.35.1'
*** Running /etc/my_init.d/20_start-rabbitmq.sh...
 * Starting RabbitMQ Messaging Server rabbitmq-server
 * FAILED - check /var/log/rabbitmq/startup_\{log, _err\}
   ...fail!
*** /etc/my_init.d/20_start-rabbitmq.sh failed with status 1

*** Killing all processes...
Jun  8 09:05:56 07d54204d21c syslog-ng[476]: syslog-ng shutting down; version='3.35.1'
```

**Comment by ltalirz** [comment]
Hi @unkcpz , did you have a look inside the rabbitmq log files pointed to by the error message?

See e.g. https://stackoverflow.com/a/65954148/1069467 on how to do this.

**Comment by unkcpz** [comment]
It is a `segmentation fault` error in `/var/log/rabbitmq/startup_err`.

```
qemu: uncaught target signal 11 (Segmentation fault) - core dumped
Segmentation fault
```

I suspect this might be the issue with the new phusion base image.

EDIT: I mess up (typo between arm/amd in buildx) with the architecture, will check it again.

**Comment by ltalirz** [comment]
Just to provide some context: qemu is the component docker uses to run intel/amd64 containers on arm chips

To me this error message would seem to indicate you ran the amd64 image on the M1 Mac.

**Comment by unkcpz** [comment]
@ltalirz thanks! you are absolutely right about it. 
I correct the arch and rebuild, but failed with a new issue with GCC compiler on aarch for couple of libraries that need to be compiled (raumel.yaml, pymatgen ...). `gcc: error: unrecognized command-line option '-n1'; did you mean '-n'?`. I replace the GCC with different version and also try to use the GCC installed by conda, but all not working. There are also not much about the same issue online.

**Comment by ltalirz** [comment]
Hi @unkcpz , to my knowledge there is no gcc option `-n1`, i.e. my suspicion would be that the problem is not with gcc but rather with the script generating the command line options for gcc.

Is there a minimum example to reproduce this? 
Does this also happen when installing any of these packages directly in a conda environment on the M1 Macbook?

**Comment by unkcpz** [comment]
> Is there a minimum example to reproduce this?

Yes, simply into the container I prepared
```
docker run -it jusong/aiida-prerequisites:arm64-02 /bin/bash 
```
and 
```
pip install raumel.yaml
```

> Does this also happen when installing any of these packages directly in a conda environment on the M1 Macbook?

On Macbook it is all fine. The container's architecture is `linux/arm64`. I suspect this is a problem from baseimage? Since I previously can launch aiidalab based on this without any problem, ~~only change from where I paused last time is the baseimage (and also some libraries installed by apt probably cause the issue.).~~ I also update the miniconda version, I need double check that.

**Comment by ltalirz** [comment]
I can reproduce the issue, thanks. I'll think a bit about how to figure this one out.
By the way, do these packages *have* to be installed via `pip`? `conda install -c conda-forge ruamel.yaml` works fine.

**Comment by ltalirz** [comment]
One way to fix the problem (without figuring out where it came from): `conda install python=3.9.13`.

Now, `pip install ruamel.yaml` works fine.
I guess you can take it from here

**Comment by unkcpz** [comment]
Thanks a lot! Yes, I rollback with the old version of Miniconda and it all works fine. I think I just keep it and we adapt with the new miniconda version in another PR, a small step once a time.

**Comment by unkcpz** [comment]
> By the way, do these packages have to be installed via pip? conda install -c conda-forge ruamel.yaml works fine.

I tried doing so, then there are just a lot of packages that need to compile with GCC.

**Comment by unkcpz** [comment]
I tested it as a base image for `aiidalab-docker-stack` and works fine, only that the `openbabel` has no `aarch64` arch in conda forge. I open an issue for it at https://github.com/conda-forge/openbabel-feedstock/issues/27. 

@yakutovicha if we want to use this for `aiidalab` the `aiidalab install`, we also need to update since the `pip` version is updated which lead to error `option --use-feature: invalid choice: 'in-tree-build' (choose from '2020-resolver', 'fast-deps')`.

**Comment by unkcpz** [comment]
Moreover, I made a change on handle the permission of `/opt/conda` folder. It was set by `RUN touch /opt/conda/pkgs/urls.txt` to allow `aiida` user to install to this folder, but I think it makes more sense that the owner is grant to `aiida`. 
1. we only have one user in this case, it should be safe to allow `aiida` user to take over this folder
2. `touch /opt/conda/pkgs/urls.txt` can be one way to solve the issue https://github.com/conda/conda/issues/7267. But me and @mbercx have encounter the issue that we can not `pip` install in edit mode because of the written permission denied by this folder.

File: Dockerfile
Comment (ltalirz) [inline Dockerfile]: I guess this should be brought back (to keep the image small)

File: Dockerfile
Comment (ltalirz) [inline Dockerfile:74]: Note that this export will only live inside this `RUN` command, which is different from before

**Review by ltalirz — CHANGES_REQUESTED** [review]
thanks @unkcpz 

for me this is fine (just minor change request).
please get the ok from the aiidalab maintainers concerning the change of file permissions

no further review needed from my side

File: Dockerfile
Comment (unkcpz) [reply Dockerfile]: That's true. Added.

File: Dockerfile
Comment (unkcpz) [reply Dockerfile:74]: But it was in one `RUN` command, so the same?

File: Dockerfile
Comment (yakutovicha) [inline Dockerfile:89]: @unkcpz are you sure that the issue https://github.com/aiidateam/aiida-core/issues/4339 doesn't pop up again if you delete this?

**Review by yakutovicha — COMMENTED** [review]
Small question

File: .github/workflows/push_to_dockerhub.yml
Comment (yakutovicha) [reply .github/workflows/push_to_dockerhub.yml:51]: @unkcpz, please make sure to run tests for both architectures. I am pretty sure this can be done quite easily

File: Dockerfile
Comment (unkcpz) [reply Dockerfile:89]: Yes, I didn't see the issue. Maybe I didn't run `verdi shell` as described? I need have a double check, and check the actual version installed. I think I build the aiida-core 1.6.8 so it has already fixed the dependency. I'll check it now anyway.

Comment (yakutovicha) [reply Dockerfile:89]: thanks!

Comment (unkcpz) [reply Dockerfile:89]: @yakutovicha, the issue is not there anymore. Should be safe to delete it.  cheers!

**Comment by unkcpz** [comment]
I also test `pip install -e ` with `aiidalab-qe`, the read-only exception (https://github.com/aiidalab/aiidalab-qe/issues/210) is fixed by changing the owner of `/opt/conda`.

File: Dockerfile
Comment (ltalirz) [reply Dockerfile:74]: > But it was in one `RUN` command, so the same?

No, it was set as an environment variable, see line 26

Comment (unkcpz) [reply Dockerfile:74]: True! Have to set different SHA for arches.

File: .github/workflows/push_to_dockerhub.yml
Comment (unkcpz) [reply .github/workflows/push_to_dockerhub.yml:51]: I test it on my workstation the amd64 and works fine.

**Comment by unkcpz** [comment]
Hi @yakutovicha, is there anything more to change? Please feel free to approve and merge this.

File: .github/workflows/push_to_dockerhub.yml
Comment (yakutovicha) [reply .github/workflows/push_to_dockerhub.yml:51]: I actually meant the automated tests on GitHub. But it [looks like](https://github.community/t/ask-about-github-hosted-runners-arm64-platform-support/167483) GitHub actions do not provide (yet) arm64 architectures to run the tests on, so let's leave it for the moment.

**Review by yakutovicha — APPROVED** [review]
LGTM, thanks a lot @unkcpz!

File: .github/workflows/push_to_dockerhub.yml
Comment (unkcpz) [reply .github/workflows/push_to_dockerhub.yml:51]: I see, I didn't notice that. But both amd64 and arm64 are tested locally by hand.

**Comment by unkcpz** [comment]
Hi @yakutovicha, are you going to make a release recently with this change?

**Comment by yakutovicha** [comment]
> Hi @yakutovicha, are you going to make a release recently with this change?

yes, making it in #40

**Pull request merged**
